homeassistant/components/tplink_omada/switch.py

```diff
@@ -51,37 +58,104 @@ async def async_setup_entry(
         coordinator = controller.get_switch_port_coordinator(switch)
         await coordinator.async_request_refresh()
 
-        for idx, port_id in enumerate(coordinator.data):
-            if idx < switch.device_capabilities.poe_ports:
-                entities.append(
-                    OmadaNetworkSwitchPortPoEControl(coordinator, switch, port_id)
-                )
+        entities.extend(
+            OmadaDevicePortSwitchEntity[
+                OmadaSwitchPortCoordinator, OmadaSwitch, OmadaSwitchPortDetails
+            ](coordinator, switch, port.port_id, _get_switch_port_base_name(port), desc)
+            for port in coordinator.data.values()
+            for desc in SWITCH_PORT_DETAILS_SWITCHES
+            if desc.exists_func(switch, port)
+        )
 
     gateway_coordinator = await controller.get_gateway_coordinator()
     if gateway_coordinator:
         for gateway in gateway_coordinator.data.values():
             entities.extend(
-                OmadaGatewayPortSwitchEntity(
-                    gateway_coordinator, gateway, p.port_number, desc
-                )
+                OmadaDevicePortSwitchEntity[
+                    OmadaGatewayCoordinator, OmadaGateway, OmadaGatewayPortStatus
+                ](gateway_coordinator, gateway, p.port_number, str(p.port_number), desc)
                 for p in gateway.port_status
-                for desc in GATEWAY_PORT_SWITCHES
-                if desc.exists_func(p)
+                for desc in GATEWAY_PORT_STATUS_SWITCHES
+                if desc.exists_func(gateway, p)
+            )
+            entities.extend(
+                OmadaDevicePortSwitchEntity[
+                    OmadaGatewayCoordinator, OmadaGateway, OmadaGatewayPortConfig
+                ](gateway_coordinator, gateway, p.port_number, str(p.port_number), desc)
+                for p in gateway.port_configs
+                for desc in GATEWAY_PORT_CONFIG_SWITCHES
+                if desc.exists_func(gateway, p)
             )
 
     async_add_entities(entities)
 
 
+def _get_switch_port_base_name(port: OmadaSwitchPortDetails) -> str:
+    """Get display name for a switch port."""
+
+    if port.name == f"Port{port.port}":
+        return f"{port.port}"
+    return f"{port.port} ({port.name})"
```

Comment:
To understand it better: Can a user rename a port to a custom name?

Reply:
For network switches, yes, ports can be renamed in the Omada controller. For gateways, no, the ports have numbers and fixed names like "WAN/LAN" that aren't visible via the API. Just another inconsistency in the Omada ecosystem. Gateways and Switches have a wildly different API behind the scenes.